### PR TITLE
plugins: move the existing plugin to a new package

### PR DESCRIPTION
**Pull request by sergiusens**
Backwards compatibility is kept for plugins that have been moved by
adding a PathFinder and Loader implementation to sys.meta_path. This
allows warnings to also be implemented for _custom_ plugin authors.

Two ensure local plugin using older import paths still work tests have
been migrated to spread with two variants with the:

- BasePlugin.
- NilPlugin

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: todo.org [hunk 1]
```diff
@@ -4,9 +4,9 @@
 
 * Specifications
 ** 2020
-*** TODO [[file:specifications/20200304-core20-plugins.org][Snapcraft Core20 Plugins]] [/]
-- [ ] Move =BasePlugin= to a legacy import path
-- [ ] Add backwards compatibility for =BasePlugin=
+*** STRT [[file:specifications/20200304-core20-plugins.org][Snapcraft Core20 Plugins]] [/]
```
Comment (cjp256) [inline todo.org]: I like the updated spec :D

File: tests/unit/plugins/v1/test_waf.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
 #
-# Copyright (C) 2016-2018 Canonical Ltd
+# Copyright (C) 2016-2018, 2020 Canonical Ltd
```
Comment (cjp256) [inline tests/unit/plugins/v1/test_waf.py]: I see a mixed use of "," and "-".  Do you prefer `2018, 2020` or `2016-2020`?  Is your editor auto-updating it?

Not asking to change it, just curious

File: snapcraft/_legacy_loader.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return module
+
+
+class LegacyPluginPathFinder(importlib.machinery.PathFinder):
+    @classmethod
+    def find_spec(cls, fullname, path=None, target=None):
+        # Ensure plugins using their original import paths can be found and
+        # warn about their new import path.
+        if fullname in [f"snapcraft.plugins.{p}" for p in _VALID_V1_PLUGINS]:
+            warnings.warn("Plugin import path has changed to 'snapcraft.plugins.v1'")
```
Comment (cjp256) [inline snapcraft/_legacy_loader.py]: nice

File: tests/unit/plugins/v1/test_waf.py (hunk 2 above)
Comment (sergiusens) [reply tests/unit/plugins/v1/test_waf.py]: Oh, 2016-2018 implies 2017 in the middle, for the cases where there were no changes in some years I resorted to `,`. And yes, I retroactively fixed the dates.

File: snapcraft/_legacy_loader.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    def exec_module(cls, module):
+        # Rewrite the module __name__ to have that of the legacy import path.
+        plugin_name = module.__name__.split(".")[-1]
+        module.__name__ = f"snapcraft.plugins.{plugin_name}"
+        return module
+
+
+class LegacyPluginPathFinder(importlib.machinery.PathFinder):
+    @classmethod
+    def find_spec(cls, fullname, path=None, target=None):
```
Comment (cjp256) [inline snapcraft/_legacy_loader.py]: types?

Comment (sergiusens) [reply snapcraft/_legacy_loader.py]: I welcome you to try, there seems to be missing mypy/typeshed bits for this to work correctly, as soon as I add any type info I get:

```
snapcraft/_legacy_loader.py:74: error: Signature of "find_spec" incompatible with supertype "MetaPathFinder"
```
Which is correct, I implemented the more lenient `PathFinder`.

``
snapcraft/_legacy_loader.py:79: error: Argument 2 to "ModuleSpec" has incompatible type "Type[LegacyPluginLoader]"; expected "Optional[Loader]"
```
Well, yeah

**Pull request merged**